File: LeetCode/91-Decode_Ways.java
/*
91. Decode Ways

*/
class Solution {
  public int numDecodings(String s) {
    if (s == null || s.length() == 0)   return 0;
    return decode(0, s);
  }

  private int decode(int index, String s) {
    if (index == s.length())    return 1;

    if (s.charAt(index) == '0') return 0;

    int count = decode(index + 1, s);
      if (index + 2 <= s.length() && Integer.parseInt(s.substring(index, index+2)) <= 26) {
       count += decode(index+2, s);
     }

    return count;
  }
}

File: LeetCode/91-Decode_Ways.py
'''
91. Decode Ways

'''
class Solution:
  def numDecodings(self, s: str) -> int:
    if not s:  return 0
    return self.helper(0, s)

  def helper(self, index, s) -> int:
    if index == len(s): return 1

    if s[index] == '0': return 0

    if index == len(s)-1: return 1



    ans = self.helper(index+1, s) \
        + (self.helper(index+2, s) if (int(s[index : index+2]) <= 26) else 0)
    return ans

